Check fixed args number for variadic function

tiif opened the pull request:
cc #4013 

@rustbot ready

oli-obk reviewed (approved):
Please squash

RalfJung commented on src/shims/unix/foreign_items.rs:
```diff
@@ -236,6 +237,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             "open" | "open64" => {
                 // `open` is variadic, the third argument is only present when the second argument has O_CREAT (or on linux O_TMPFILE, but miri doesn't support that) set
                 this.check_abi_and_shim_symbol_clash(abi, Conv::C , link_name)?;
+                this.check_fixed_args_count("open/open64", abi, 2)?;
```
Now it's still fairly easy to forget to call `check_fixed_args_count`.

So instead, please add a new `this.check_shim_variadic` that is a lot like `check_shim` but returns something like `(&'a [OpTy<'tcx>; N], &'a [OpTy<'tcx>])` where the "fixed" arguments are mapped to the array and the variadic ones are returned in the slice.

RalfJung commented on src/helpers.rs:
```diff
@@ -920,6 +920,22 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+    ) -> InterpResult<'tcx> {
+        if abi.fixed_count != expected_count {
+            throw_ub_format!(
+                "incorrect number of fixed args for `{name}`: got {}, expected {expected_count}",
```
```suggestion
                "incorrect number of fixed arguments for variadic function `{name}`: got {}, expected {expected_count}",
```

tiif commented:
@rustbot author

tiif commented:
So now we have this weird function definition
```rust
    fn open(
        &mut self,
        args: &[OpTy<'tcx>],
        fixed: &[OpTy<'tcx>; 2],
        _var: &[OpTy<'tcx>],
    ) -> InterpResult<'tcx, Scalar> {

```

where ``args`` is the full function argument, ``fixed`` is the fixed args, and ``_var`` is var args. 

Since we already split fixed and var args through ``check_shim_variadic`` before we call ``this.open``, I have tried to not pass ``args`` into ``open``, but the problem is if I avoid using ``args`` and do this instead

```
            let [mode] = check_min_arg_count("open(pathname, O_CREAT, ...)", var)?
```
It will lead to diagnostic like this:
```rust
11 |     let _fd = unsafe { libc::open(name_ptr, libc::O_CREAT) }; 
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ incorrect number of arguments for `open(pathname, O_CREAT, ...)`: got 0, expected at least 1
```

Or maybe we can just ignore the fixed args and var args returned by ``check_shim_variadic``?

RalfJung commented:
> It will lead to diagnostic like this:

I would suggest to rename `check_min_arg_count` to `check_min_vararg_count` and make the error say something like "not enough variadic arguments for ...".

oli-obk commented on src/shims/unix/linux_like/syscall.rs:
```diff
@@ -15,8 +15,8 @@ pub fn syscall<'tcx>(
-    // count is checked bellow.
-    ecx.check_abi_and_shim_symbol_clash(abi, Conv::C, link_name)?;
+    // count is checked below.
+    let ([op], _var) = ecx.check_shim_variadic::<1>(abi, Conv::C, link_name, "syscall", args)?;
```
```suggestion
    let ([op], _var) = ecx.check_shim_variadic(abi, Conv::C, link_name, "syscall", args)?;
```

is that explicit length necessary? doesn't it get inferred from the return type's array length?

tiif replied:
Oh yea, it does

tiif commented on src/shims/unix/foreign_items.rs:
```diff
@@ -207,8 +207,9 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             "fcntl" => {
                 // `fcntl` is variadic. The argument count is checked based on the first argument
```
This comment looks weird, the argument count should be checked based the second argument ``cmd`` instead the first? ([source](https://docs.rs/libc/latest/libc/fn.fcntl.html))

RalfJung replied:
Probably a copy-paste mistake.

tiif commented:
@rustbot ready

RalfJung commented:
> I think as a miri specific extern function, the argument here shouldn't be classified as a vararg?

Oh dang.

We have a comment in the docs for this function saying "The `flags` argument must be `1`."
So could you try making a PR that forces the function to have 2 arguments and `flags == 1`? We recently did the same with `miri_resolve_frame`, removing support for the old "version 0 protocol"; if we do the same here then we can get rid of this use of `check_min_arg_count` hopefully. :)

RalfJung commented on src/helpers.rs:
```diff
@@ -928,6 +928,30 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         interp_ok(())
     }
 
+    /// Check the number of fixed args.
```
```suggestion
    /// Check the number of fixed args of a vararg function.
```
This should probably error if the function is *not* a vararg fn?

Please also put "vararg" in the name then, e.g. `check_vargarg_fixed_arg_count`.

(We used `arg_count` before so please don't use `args_count` here now.)

RalfJung commented on src/helpers.rs:
```diff
@@ -990,6 +1014,23 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+        &'a [OpTy<'tcx>; N]: TryFrom<&'a [OpTy<'tcx>]>,
+    {
+        self.check_abi_and_shim_symbol_clash(abi, exp_abi, link_name)?;
+        self.check_fixed_args_count(shim_name, abi, args)
```
Why does this need `self` when `check_arg_count` does not?

tiif replied:
Oh ``check_arg_count`` is free function, I will move ``check_vargarg_fixed_arg_count`` to free function too.

RalfJung commented on src/helpers.rs:
```diff
@@ -1183,7 +1224,7 @@ where
 }
 
-/// Check that the number of args is at least the minumim what we expect.
+/// Check that the number of args is at least the minimum what we expect.
```
```suggestion
/// Check that the number of args is at least the minimum what we expect.
/// FIXME: Remove this function, use varargs and `check_min_vararg_count` instead.
```

RalfJung commented on src/shims/unix/android/thread.rs:
```diff
@@ -16,18 +16,18 @@ pub fn prctl<'tcx>(
-    // We do not use `check_shim` here because `prctl` is variadic. The argument
-    // count is checked bellow.
-    ecx.check_abi_and_shim_symbol_clash(abi, Conv::C, link_name)?;
+    // `prctl` is variadic. The argument count is checked below.
```
```suggestion
```
Not sure which value this comment still adds.

RalfJung commented on src/shims/unix/android/thread.rs:
```diff
@@ -16,18 +16,18 @@ pub fn prctl<'tcx>(
     let pr_get_name = 16;
 
-    let [op] = check_min_arg_count("prctl", args)?;
+    let [op] = fixed_args;
```
```suggestion
```

RalfJung commented on src/shims/unix/android/thread.rs:
```diff
@@ -16,18 +16,18 @@ pub fn prctl<'tcx>(
-    // count is checked bellow.
-    ecx.check_abi_and_shim_symbol_clash(abi, Conv::C, link_name)?;
+    // `prctl` is variadic. The argument count is checked below.
+    let (fixed_args, varargs) =
+        ecx.check_shim_variadic::<1>(abi, Conv::C, link_name, "prctl", args)?;
```
```suggestion
    let ([op], varargs) =
        ecx.check_shim_variadic(abi, Conv::C, link_name, "prctl", args)?;
```
Same everywhere else -- you should never use turbofish here.

RalfJung commented on src/shims/unix/fd.rs:
```diff
@@ -127,10 +127,14 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
-    fn fcntl(&mut self, args: &[OpTy<'tcx>]) -> InterpResult<'tcx, Scalar> {
+    fn fcntl(
+        &mut self,
+        fixed_args: &[OpTy<'tcx>; 2],
```
Instead of passing a 2-element array, just pass two arguments.

RalfJung reviewed (commented):
I left some comments on the first place where you used the new infra; they apply everywhere in the PR.

RalfJung commented:
@rustbot author

tiif commented:
> So could you try making a PR that forces the function to have 2 arguments and flags == 1? We recently did the same with miri_resolve_frame, removing support for the old "version 0 protocol"; if we do the same here then we can get rid of this use of check_min_arg_count hopefully. :)

Alright, I can send a follow-up PR after this landed.

tiif commented on src/helpers.rs:
```diff
@@ -1198,6 +1216,49 @@ pub fn check_min_arg_count<'a, 'tcx, const N: usize>(
+    abi: &FnAbi<'tcx, Ty<'tcx>>,
+    args: &'a [OpTy<'tcx>],
+) -> InterpResult<'tcx, (&'a [OpTy<'tcx>; N], &'a [OpTy<'tcx>])> {
+    if !abi.c_variadic {
+        panic!("This should not be used on non-vararg function");
+    }
```
I am going for a panic here, but we can use other mechanism to prevent this being used at the wrong place.

RalfJung replied:
Which other mechanism did you have in mind?

RalfJung replied:
Actually yes this should not be a panic, this should be `throw_ub_format!`. It is UB to call a variadic function with a non-variadic caller-side signature.

tiif replied:
> t is UB to call a variadic function with a non-variadic caller-side signature.

I wonder if there is any way to test for this?

RalfJung replied:
Yes, just write an `extern` block declaring `open` as a non-variadic function and call that.

tiif commented:
@rustbot ready

RalfJung commented on src/shims/unix/foreign_items.rs:
```diff
@@ -236,8 +235,9 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
-                this.check_abi_and_shim_symbol_clash(abi, Conv::C, link_name)?;
-                let result = this.open(args)?;
+                let ([path_raw, flag], varargs) =
+                    this.check_shim_variadic(abi, Conv::C, link_name, "open/open64", args)?;
```
Just keep using `link_name` here, that will always match the name the user used which is better. This also applies everywhere else you changed this.

RalfJung commented:
@rustbot author

RalfJung commented on tests/fail-dep/libc/wrong_fixed_arg_count.rs:
This should not be a pass-dep test, it's not actually about libc after all.  `tests/fail/shims` seems like a better place.

tiif replied:
After I moved it to ``./tests/fail/shims``, when I run ``./miri test``, the error ``undeclared crate or module `libc` `` appears. 

```rust
error[E0433]: failed to resolve: use of undeclared crate or module `libc`
  --> tests/fail/shims/wrong_fixed_arg_count.rs:12:46
   |
LL |     let _fd = unsafe { open(c_path.as_ptr(), libc::O_RDWR) };
   |                                              ^^^^ use of undeclared crate or module `libc`

error[E0433]: failed to resolve: use of undeclared crate or module `libc`
  --> tests/fail/shims/wrong_fixed_arg_count.rs:7:26
   |
LL |     fn open(path: *const libc::c_char, ...) -> libc::c_int;
   |                          ^^^^ use of undeclared crate or module `libc`

error[E0433]: failed to resolve: use of undeclared crate or module `libc`
  --> tests/fail/shims/wrong_fixed_arg_count.rs:7:48
   |
LL |     fn open(path: *const libc::c_char, ...) -> libc::c_int;
   |                                                ^^^^ use of undeclared crate or module `libc`

```

I guess if I need to use something related to ``libc``, I need to put it in ``fail-dep``.

tiif replied:
Hmm should this even happen? The comment inside test_dependencies said ``libc`` should be able to be used in all files under ``tests/``

https://github.com/rust-lang/miri/blob/ce9c9ff2de0aaf410fd461d4f8349b8765894820/test_dependencies/Cargo.toml#L11-L12

RalfJung replied:
Replace the use of `libc` by [`std:ffi`](https://doc.rust-lang.org/nightly/std/ffi/index.html).

RalfJung replied:
> The comment inside test_dependencies said libc should be able to be used in all files under tests/

The comment is wrong; it is only for the tests in `*-deps`.

RalfJung commented on src/helpers.rs:
```diff
@@ -1198,6 +1215,53 @@ pub fn check_min_arg_count<'a, 'tcx, const N: usize>(
+) -> InterpResult<'tcx, (&'a [OpTy<'tcx>; N], &'a [OpTy<'tcx>])> {
+    if !abi.c_variadic {
+        throw_ub_format!(
+            "it is incorrect to call variadic function with a non-variadic caller-side signature"
```
```suggestion
            "calling a variadic function with a non-variadic caller-side signature"
```

tiif commented:
@rustbot ready

RalfJung commented:
Please also squash the PR into a single commit.

RalfJung commented:
@rustbot author

tiif commented:
Ooh I accidentally push to github then rebase then push, so the change is not visible from ``compare change``. 

What has changed:
- I use ``/* value does not matter */ 0`` for ``wrong_fixed_arg_count ``test too since it also has magic number
- bless the test 

@rustbot ready

RalfJung commented:
Looks good :)
